README.md
<div align="center">

![react-promise-loader](https://repository-images.githubusercontent.com/236146663/28c41100-3f90-11ea-9922-a1ed44b60cbe)

[![Version](https://img.shields.io/npm/v/react-promise-loader)](https://www.npmjs.com/package/react-promise-loader)
[![Build](https://img.shields.io/github/workflow/status/awibox/react-promise-loader/Deploy)](https://github.com/awibox/react-promise-loader/actions)
[![Coverage](https://coveralls.io/repos/github/awibox/react-promise-loader/badge.svg?branch=master)](https://coveralls.io/github/awibox/react-promise-loader?branch=master)
[![Minified size](https://img.shields.io/bundlephobia/min/react-promise-loader)](https://github.com/awibox/react-promise-loader/blob/master/LICENSE)
[![Downloads](https://img.shields.io/npm/dm/react-promise-loader)](https://www.npmjs.com/package/react-promise-loader)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)](https://github.com/awibox/react-promise-loader/pulls)
[![Tested with jest](https://img.shields.io/badge/tested_with-jest-99424f.svg)](https://github.com/facebook/jest)

</div>

## Table of contents
* [Installation](#installation)
* [Getting started](#gettingstarted)
* [The settings of the component](#settings)
* [Contributing](#contributing)

<a name="installation"></a>
## Installation
You need to install promise loader and promise tracker packages:
```bash
npm install react-promise-loader react-promise-tracker
```
You can use yarn:
```bash
yarn add react-promise-loader react-promise-tracker
```
<a name="gettingstarted"></a>
## Getting started
Wrap promises that you would like to track as in the following example:
```diff
+ import { trackPromise} from 'react-promise-tracker';
//...

+ trackPromise(
    fetchUsers(); // You function that returns a promise
+ );
```
Then you need to add the Loader component and send data from the react-promise-tracker:
```diff
+ import { usePromiseTracker } from 'react-promise-tracker';
+ import Loader from 'react-promise-loader';

const App = () => (
  <div className="app">
    ...
+     <Loader promiseTracker={usePromiseTracker} color={'#3d5e61'} background={'rgba(255,255,255,.5)'} />
  </div>
);
export default App;
```

<a name="settings"></a>
## The settings of the component
| Parameter          | Type    | Description                                                                           | Default                      |
|--------------------|---------|---------------------------------------------------------------------------------------|------------------------------|
| **background**     | string  | Sets the color for the background in any format that supports css                     | ``` rgba(255,255,255,.5) ``` |
| **color**          | string  | Sets the color of the spinner                                                         | ``` #000 ```                 |
| **promiseTracker** | boolean | You need to set ```usePromiseTracker``` function from the ```react-promise-tracker``` | ``` false ```                |
| **loading**        | boolean | If you need to run the loader without tracking promises you should set ```true```     | ``` false ```                |
| **zIndex**         | number  | You can change the z-index to distribute the layers correctly                         | ``` 999 ```                  |

<a name="contributing"></a>
## Contributing
Please read through our [CONTRIBUTING.md](/.github/CONTRIBUTING.md).


CHANGELOG.md
# [1.2.0](https://github.com/awibox/react-promise-loader/compare/v1.1.4...v1.2.0) (2022-06-28)


### Features

* **zIndex:** [#321](https://github.com/awibox/react-promise-loader/issues/321) ability to add custom z-index ([#367](https://github.com/awibox/react-promise-loader/issues/367)) ([65a2fee](https://github.com/awibox/react-promise-loader/commit/65a2feecd7bd654f13a861ba4632dd400d0726bc))

## [1.1.4](https://github.com/awibox/react-promise-loader/compare/v1.1.3...v1.1.4) (2022-06-28)


### Bug Fixes

* **dependencies:** dependencies were updated ([2b11662](https://github.com/awibox/react-promise-loader/commit/2b116620577831d35fc9da1d79b824c0a5f9f2ae))

# 1.0.0 (2022-06-28)


### Bug Fixes

* git ign ([c5e558e](https://github.com/awibox/react-promise-loader/commit/c5e558e03e8b8a3a4f10f2f17c9c46514cb0a899))
* switch to dist dir ([022861a](https://github.com/awibox/react-promise-loader/commit/022861ac08423330378fee4ddc6fa7e414bd575c))
* Update dependencies ([a18c6c1](https://github.com/awibox/react-promise-loader/commit/a18c6c1781ce8bf1b8d40129f595de75d111da5d))
* update package.json ([0f69283](https://github.com/awibox/react-promise-loader/commit/0f69283f9fa56d96476fec67705aca9ec858e927))


### Features

* semantic relese ([cc6acd2](https://github.com/awibox/react-promise-loader/commit/cc6acd215beb604e3c899267d7d9274fe9e9d399))

# 1.0.0 (2022-06-28)


### Bug Fixes

* git ign ([c5e558e](https://github.com/awibox/react-promise-loader/commit/c5e558e03e8b8a3a4f10f2f17c9c46514cb0a899))
* switch to dist dir ([022861a](https://github.com/awibox/react-promise-loader/commit/022861ac08423330378fee4ddc6fa7e414bd575c))
* Update dependencies ([a18c6c1](https://github.com/awibox/react-promise-loader/commit/a18c6c1781ce8bf1b8d40129f595de75d111da5d))
* update package.json ([0f69283](https://github.com/awibox/react-promise-loader/commit/0f69283f9fa56d96476fec67705aca9ec858e927))


### Features

* semantic relese ([cc6acd2](https://github.com/awibox/react-promise-loader/commit/cc6acd215beb604e3c899267d7d9274fe9e9d399))

# 1.0.0 (2022-06-25)


### Bug Fixes

* git ign ([c5e558e](https://github.com/awibox/react-promise-loader/commit/c5e558e03e8b8a3a4f10f2f17c9c46514cb0a899))
* switch to dist dir ([022861a](https://github.com/awibox/react-promise-loader/commit/022861ac08423330378fee4ddc6fa7e414bd575c))
* Update dependencies ([a18c6c1](https://github.com/awibox/react-promise-loader/commit/a18c6c1781ce8bf1b8d40129f595de75d111da5d))
* update package.json ([0f69283](https://github.com/awibox/react-promise-loader/commit/0f69283f9fa56d96476fec67705aca9ec858e927))


### Features

* semantic relese ([cc6acd2](https://github.com/awibox/react-promise-loader/commit/cc6acd215beb604e3c899267d7d9274fe9e9d399))

# 1.0.0 (2022-06-24)


### Bug Fixes

* git ign ([c5e558e](https://github.com/awibox/react-promise-loader/commit/c5e558e03e8b8a3a4f10f2f17c9c46514cb0a899))
* switch to dist dir ([022861a](https://github.com/awibox/react-promise-loader/commit/022861ac08423330378fee4ddc6fa7e414bd575c))
* Update dependencies ([a18c6c1](https://github.com/awibox/react-promise-loader/commit/a18c6c1781ce8bf1b8d40129f595de75d111da5d))
* update package.json ([0f69283](https://github.com/awibox/react-promise-loader/commit/0f69283f9fa56d96476fec67705aca9ec858e927))


### Features

* semantic relese ([cc6acd2](https://github.com/awibox/react-promise-loader/commit/cc6acd215beb604e3c899267d7d9274fe9e9d399))

# 1.0.0 (2022-06-24)


### Bug Fixes

* git ign ([c5e558e](https://github.com/awibox/react-promise-loader/commit/c5e558e03e8b8a3a4f10f2f17c9c46514cb0a899))
* switch to dist dir ([022861a](https://github.com/awibox/react-promise-loader/commit/022861ac08423330378fee4ddc6fa7e414bd575c))
* Update dependencies ([a18c6c1](https://github.com/awibox/react-promise-loader/commit/a18c6c1781ce8bf1b8d40129f595de75d111da5d))
* update package.json ([0f69283](https://github.com/awibox/react-promise-loader/commit/0f69283f9fa56d96476fec67705aca9ec858e927))


### Features

* semantic relese ([cc6acd2](https://github.com/awibox/react-promise-loader/commit/cc6acd215beb604e3c899267d7d9274fe9e9d399))

# 1.0.0 (2022-06-24)


### Bug Fixes

* git ign ([c5e558e](https://github.com/awibox/react-promise-loader/commit/c5e558e03e8b8a3a4f10f2f17c9c46514cb0a899))
* switch to dist dir ([022861a](https://github.com/awibox/react-promise-loader/commit/022861ac08423330378fee4ddc6fa7e414bd575c))
* Update dependencies ([a18c6c1](https://github.com/awibox/react-promise-loader/commit/a18c6c1781ce8bf1b8d40129f595de75d111da5d))
* update package.json ([0f69283](https://github.com/awibox/react-promise-loader/commit/0f69283f9fa56d96476fec67705aca9ec858e927))


### Features

* semantic relese ([cc6acd2](https://github.com/awibox/react-promise-loader/commit/cc6acd215beb604e3c899267d7d9274fe9e9d399))

# 1.0.0 (2022-06-23)


### Bug Fixes

* git ign ([c5e558e](https://github.com/awibox/react-promise-loader/commit/c5e558e03e8b8a3a4f10f2f17c9c46514cb0a899))
* switch to dist dir ([022861a](https://github.com/awibox/react-promise-loader/commit/022861ac08423330378fee4ddc6fa7e414bd575c))
* Update dependencies ([a18c6c1](https://github.com/awibox/react-promise-loader/commit/a18c6c1781ce8bf1b8d40129f595de75d111da5d))
* update package.json ([0f69283](https://github.com/awibox/react-promise-loader/commit/0f69283f9fa56d96476fec67705aca9ec858e927))


### Features

* semantic relese ([cc6acd2](https://github.com/awibox/react-promise-loader/commit/cc6acd215beb604e3c899267d7d9274fe9e9d399))

# 1.0.0 (2022-06-23)


### Bug Fixes

* git ign ([c5e558e](https://github.com/awibox/react-promise-loader/commit/c5e558e03e8b8a3a4f10f2f17c9c46514cb0a899))
* switch to dist dir ([022861a](https://github.com/awibox/react-promise-loader/commit/022861ac08423330378fee4ddc6fa7e414bd575c))
* Update dependencies ([a18c6c1](https://github.com/awibox/react-promise-loader/commit/a18c6c1781ce8bf1b8d40129f595de75d111da5d))

jest-setup-tests.js
import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { createSerializer } from 'enzyme-to-json';

// Set the default serializer for Jest to be the from enzyme-to-json
// This produces an easier to read (for humans) serialized format.
expect.addSnapshotSerializer(createSerializer({ mode: 'deep' }));

// React 16 Enzyme adapter
Enzyme.configure({ adapter: new Adapter() });

// Define globals to cut down on imports in test files
global.React = React;
global.shallow = shallow;

// Use fake timers for async events
jest.useFakeTimers();


specs/Loader.spec.js
import React from 'react';
import { shallow } from 'enzyme';
import Loader from '../src/components/Loader';

describe('Loader', () => {
  it('should render correctly', () => {
    const component = shallow(<Loader loading={true} />);
    expect(component).toMatchSnapshot();
  });
  it('should not render', () => {
    const component = shallow(<Loader />);
    expect(component.get(0)).toEqual(null);
  });
  it('should render for promiseTracker', () => {
    const component = shallow(<Loader promiseTracker={() => ({ promiseInProgress: true })} />);
    expect(component).toExist();
  });
});


src/components/Loader.js
import React from 'react';
import PropTypes from 'prop-types';

const Loader = (props) => {
  const {
    color, background, loading, promiseTracker, zIndex,
  } = props;

  const defaultLoaderStyle = {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    zIndex,
    background: 'rgba(255,255,255,.5)',
    backfaceVisibility: 'hidden',
  };

  const defaultSpinnerStyle = {
    position: 'absolute',
    left: '50%',
    top: '50%',
    transform: 'translate(-50%, -50%)',
    width: 100,
    height: 100,
  };

  const { promiseInProgress } = promiseTracker ? promiseTracker() : false;
  const loaderStyle = {
    ...defaultLoaderStyle,
    background,
  };
  const spinnerItemsArray = [
    { transform: 'rotate(0 50 50)', begin: '-0.9166666666666666s' },
    { transform: 'rotate(30 50 50)', begin: '-0.8333333333333334s' },
    { transform: 'rotate(60 50 50)', begin: '-0.75s' },
    { transform: 'rotate(90 50 50)', begin: '-0.6666666666666666s' },
    { transform: 'rotate(120 50 50)', begin: '-0.5833333333333334s' },
    { transform: 'rotate(150 50 50)', begin: '-0.5s' },
    { transform: 'rotate(180 50 50)', begin: '-0.4166666666666667s' },
    { transform: 'rotate(210 50 50)', begin: '-0.3333333333333333s' },
    { transform: 'rotate(240 50 50)', begin: '-0.25s' },
    { transform: 'rotate(270 50 50)', begin: '-0.16666666666666666s' },
    { transform: 'rotate(300 50 50)', begin: '-0.08333333333333333s' },
    { transform: 'rotate(330 50 50)', begin: '0s' },
  ];
  return (
    loading || promiseInProgress
      ? <div style={loaderStyle}>
      <div style={defaultSpinnerStyle}>
        <svg className="lds-spinner" width="100%" height="100%" xmlns="http://www.w3.org/2000/svg"
             viewBox="0 0 100 100" preserveAspectRatio="xMidYMid">
          {spinnerItemsArray.map((item) => <g key={item.transform} transform={item.transform}>
            <rect x="47" y="24" rx="9.4" ry="4.8" width="6" height="12" fill={color}>
              <animate attributeName="opacity" values="1;0" keyTimes="0;1" dur="1s" begin={item.begin} repeatCount="indefinite"/>
            </rect>
          </g>)}
        </svg>
      </div>
    </div> : ''
  );
};

Loader.propTypes = {
  color: PropTypes.string,
  loading: PropTypes.bool,
  background: PropTypes.string,
  promiseTracker: PropTypes.any,
  zIndex: PropTypes.number,
};

Loader.defaultProps = {
  color: '#000',
  loading: false,
  background: 'rgba(255,255,255,.5)',
  promiseTracker: false,
  zIndex: 999,
};

export default Loader;
